### PR TITLE
Growth rate dataset convention and application

### PR DESCRIPTION
**Pull request by lixiliu**
Completes JIRA Story DSGRID-
Closes GitHub Issue # 213
Companion dsgrid-project-STandardScenarios PR: https://github.com/dsgrid/dsgrid-project-StandardScenarios/pull/83

Refactor code base to allow for new projection dataset construction methods.

ResStock and ComStock projection datasets now have the following config:
dataset_type: "projection" (previously "exponential_growth")
construction_method: "exponential_growth" (previously "formula123")

DatasetType.PROJECTION can now have various DatasetConstructionMethod defined, e.g., EXPONENTIAL_GROWTH, ANNUAL_MULTIPLIER, which will have different dataset processing functions in growth_rates.py.

## Checklist

- [ ] Tests exercising the new feature or bug fix
- [ ] All tests pass
- [ ] At least one code review approval
- [ ] Consider transferring TODOs to JIRA or GitHub

### CODE REVIEW

File: dsgrid/query/models.py [hunk 1]
```diff
@@ -303,11 +303,18 @@ class ProjectQueryDatasetParamsModel(CacheableQueryBaseModel):
 class DatasetType(enum.Enum):
     """Defines the type of a dataset in a query."""
 
-    EXPONENTIAL_GROWTH = "exponential_growth"
+    PROJECTION = "projection"
     STANDALONE = "standalone"
     DERIVED = "derived"
 
 
+class DatasetConstructionMethod(enum.Enum):
```
Comment (daniel-thom) [inline dsgrid/query/models.py:311]:
```suggestion
class DatasetConstructionMethod(str, enum.Enum):
```

File: dsgrid/query/models.py [hunk 2]
```diff
@@ -331,22 +338,23 @@ def get_dataset_id(self) -> str:
...
 
 
-class ExponentialGrowthDatasetModel(DatasetBaseModel):
+class ProjectionDatasetModel(DatasetBaseModel):
     """A dataset with growth rates that can be applied to a standalone dataset."""
 
     dataset_type: Annotated[
-        Literal[DatasetType.EXPONENTIAL_GROWTH.value],
-        Field(default=DatasetType.EXPONENTIAL_GROWTH.value),
+        Literal[DatasetType.PROJECTION.value],
```
Comment (daniel-thom) [inline dsgrid/query/models.py:345]:
```suggestion
        Literal[DatasetType.PROJECTION],
```

File: dsgrid/query/models.py [hunk 3]
```diff
@@ -331,22 +338,23 @@ def get_dataset_id(self) -> str:
...
+        Field(default=DatasetType.PROJECTION.value),
```
Comment (daniel-thom) [inline dsgrid/query/models.py:346]:
```suggestion
        Field(default=DatasetType.PROJECTION),
```

File: tests/test_derived_datasets.py [hunk 4]
```diff
@@ -52,11 +52,11 @@ def valid_query():
             dataset=DatasetModel(
                 dataset_id="resstock_conus_2022_projected",
                 source_datasets=[
-                    ExponentialGrowthDatasetModel(
+                    ProjectionDatasetModel(
                         dataset_id="resstock_conus_2022_projected",
                         initial_value_dataset_id="resstock_conus_2022_reference",
                         growth_rate_dataset_id="aeo2021_reference_residential_energy_use_growth_factors",
-                        construction_method="formula123",
+                        construction_method="exponential_growth",
```
Comment (daniel-thom) [inline tests/test_derived_datasets.py:59]: Needs to be the enum, as in `DatasetConstructionMethod.EXPONENTIAL_GROWTH`

File: dsgrid/dataset/growth_rates.py [hunk 5]
```diff
@@ -34,24 +42,58 @@ def apply_growth_rate_123(
...
+    """Applies annual growth rate to the initial_value dataframe as follows:
+    P(t) = P0 * r(t)
+    where:
+      P(t): quantity at year t
+      P0: initial quantity
+      r(t): growth rate per year t (relative to P0)
+
+    Parameters
+    ----------
+    initial_value_df : pyspark.sql.DataFrame
```
Comment (daniel-thom) [inline dsgrid/dataset/growth_rates.py:79]: We are not at all consistent with these docstrings. If the functions are only going to be called internally and we want to specify the type only once, I would declare it as a type annotation in the function definition instead. No need to change the code again; this is just what I’ve been doing. We can try to standardize this.

**Review by daniel-thom — APPROVED** [review]

File: dsgrid/query/models.py [hunk 6]
```diff
@@ -300,14 +300,21 @@ class ProjectQueryDatasetParamsModel(CacheableQueryBaseModel):
...
+class DatasetType(str, enum.Enum):
     """Defines the type of a dataset in a query."""
 
-    EXPONENTIAL_GROWTH = "exponential_growth"
+    PROJECTION = "projection"
     STANDALONE = "standalone"
     DERIVED = "derived"
 
 
+class DatasetConstructionMethod(str, enum.Enum):
```
Comment (elainethale) [inline dsgrid/query/models.py:311]: nit: Could be "DatasetProjectionMethod"

**Pull request merged**